src/Connection.ts

```diff
@@ -1,1488 +1,1884 @@
+	/** Used to bypass the cache */
+	forceUpdate?: boolean;
+	/** Can be used to identify the request method in error messages */
+	requestName?: string;
+	/**
+	 * The timeout in milliseconds after which the call will reject with a timeout error.
+	 * If no timeout is given, the default is used. Set this to `false` to explicitly disable the timeout.
+	 */
+	commandTimeout?: number | false;
+	/** Will be called when the timeout elapses */
+	onTimeout?: () => void;
+	/** Whether the call should only be allowed in the admin adapter */
+	requireAdmin?: boolean;
+	/** Require certain features to be supported for this call */
+	requireFeatures?: string[];
+	/** The function that does the actual work */
+	executor: (
+		resolve: (value: T | PromiseLike<T> | Promise<T>) => void,
+		reject: (reason?: any) => void,
+		/** Can be used to check in the executor whether the request has timed out and/or stop it from timing out */
+		timeout: Readonly<{ elapsed: boolean; clearTimeout: () => void }>,
+	) => void | Promise<void>;
+}
+
+export type SystemConfig = NonNullable<
+	ioBroker.ObjectIdToObjectType<"system.config">
+>;
+export type BinaryStateChangeHandler = (
+	id: string,
+	base64: string | null,
+) => void;
+
+export interface OldObject {
+	_id: string;
+	type: string;
+}
+
+export type ObjectChangeHandler = (
+	id: string,
+	obj: ioBroker.Object | null | undefined,
+	oldObj?: OldObject,
+) => void | Promise<void>;
+
+export class Connection<
+	CustomListenEvents extends Record<
+		keyof CustomListenEvents,
+		ListenEventHandler
+	> = Record<string, never>,
+	CustomEmitEvents extends Record<
+		keyof CustomEmitEvents,
+		EmitEventHandler
+	> = Record<string, never>,
+> {
+	constructor(props: Partial<ConnectionProps>) {
+		this.props = this.applyDefaultProps(props);
+
+		this.waitForSocketLib()
+			.then(() => this.startSocket())
+			.catch((e) => {
+				alert(`Socket connection could not be initialized: ${e}`);
+			});
+	}
+
+	private applyDefaultProps(
+		props: Partial<ConnectionProps>,
+	): ConnectionProps {
+		return {
+			...props,
+			// Define default props that always need to be set
+			protocol: props.protocol || window.location.protocol,
+			host: props.host || window.location.hostname,
+			port:
+				props.port ||
+				(window.location.port === "3000" ? 8081 : window.location.port),
+			ioTimeout: Math.max(props.ioTimeout || 20000, 20000),
+			cmdTimeout: Math.max(props.cmdTimeout || 5000, 5000),
+			admin5only: props.admin5only || false,
+			autoSubscribes: props.autoSubscribes ?? [],
+			autoSubscribeLog: props.autoSubscribeLog ?? false,
+			doNotLoadACL: props.doNotLoadACL ?? true,
+			doNotLoadAllObjects: props.doNotLoadAllObjects ?? true,
+		};
+	}
+
+	private readonly props: ConnectionProps;
+
+	private connected: boolean = false;
+	private subscribed: boolean = false;
+	private firstConnect: boolean = true;
+	public waitForRestart: boolean = false;
+	public loaded: boolean = false;
+
+	private statesSubscribes: Record<
```

Comment:
Should be readonly

Reply:
... same for all `Record<...>` below

Reply:
Can't do it for `objects`. They should probably be Maps instead, but that's something for a later PR

Reply:
I was thinking about Maps as well... agree with a later PR.